README.md
Panedr
======

[![Build Status](https://travis-ci.org/jbarnoud/panedr.svg)](https://travis-ci.org/jbarnoud/panedr)

Panedr reads a Gromacs EDR file and returns its content as a pandas dataframe. The library exposes one function—the `edr_to_df` function—that gets the path to an EDR file and returns a pandas dataframe.

The pandas library is required.

Example
-------

```python
import panedr

# Read the EDR file
path = 'ener.edr'
df = panedr.edr_to_edr(path)

# Get the average pressure after the first 10 ns
pressure_avg = df[u'Pressure'][df[u'Time'] > 1000].mean()
```

Install
-------

The ``panedr`` module can be installed from source. The source files are available on Github at https://github.com/jbarnoud/panedr. Once the file downloaded, go to the panedr directory and run:

```bash
pip install .
```

You may not have the permissions to install modules system wide, or you may want to install ``panedr`` only for the current user. To install the module user wide only, run:

```bash
pip install --user .
```

You may want to consider [virtualenv](http://docs.python-guide.org/en/latest/dev/virtualenvs/).

`panedr` is compatible with Python 2.7 and greater and python 3.3 and greater.

Tests
-----

The `panedr` module comes with a series of tests. To run these tests, install [nose](https://nose.readthedocs.org/en/latest/), and run:

```bash
nosetests -v panedr
```

License
-------

Panedr translate in python part of the source code of Gromacs. Therefore, Panedr is distributed under the same GNU Lesser General Public License version 2.1 as Gromacs.

> Panedr — a library to manipulate Gromacs EDR file in python
>
> Copyright (C) 2016  Jonathan Barnoud
> 
> This library is free software; you can redistribute it and/or
> modify it under the terms of the GNU Lesser General Public
> License as published by the Free Software Foundation; either
> version 2.1 of the License, or (at your option) any later version.
> 
> This library is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
> Lesser General Public License for more details.
> 
> You should have received a copy of the GNU Lesser General Public
> License along with this library; if not, write to the Free Software
> Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA


ChangeLog
# -*- tab-width: 2; indent-tabs-mode: nil; coding: utf-8 -*-
==============================
  Pyedr and Panedr CHANGELOG
==============================

The rules for this file:
  * release numbering uses semantic versioning 2.0 https://semver.org/
  * entries are sorted newest-first.
  * summarize sets of changes - don't reproduce every git log comment here.
  * don't ever delete anything.
  * keep the format consistent (79 char width, M/D/Y date format) and do not
    use tabs but use spaces for formatting
  * accompany each entry with github issue/PR number (Issue #xyz)
  * Note: rules were not applied before v0.6.0

------------------------------------------------------------------------------
??/??/???? IAlibay

  * 0.9.0

  Changes


12/23/2023 ezavod, IAlibay

  * 0.8.0

  Changes

  * Added support for legacy file version 1 (PR #22)
  * Extended test coverage and code cleanup (PR #22)
  * Add support for Python 3.12 (PR #71 and #74)

09/04/2023 IAlibay

  * 0.7.2

  Changes

  * The mda-xdrlib library is now used instead of the deprecated xdrlib
    Python library (PR #67)

12/15/2022 BFedder, IAlibay

  * 0.7.1
  
  Changes
  
  * Stopped using time.time() in verbose reporting and switched to using
    tqdm for progress bars (PR # 55)
  * Replaced some of the test data files with smaller counterparts to reduce
    total file size (PR #55)
  * Added Python 3.11 support (PR #59)

  Fixes

  * Adds pyedr/tests/data to manifest for packaging.


02/09/2022 BFedder, IAlibay

  * 0.7.0
  
  Changes
  
  * Add a new `get_unit_dictionary` method which returns a dictionary
    with the units of each energy term found in the EDR file (PR #56)


15/06/2022 jbarnoud, BFedder, orbeckst, hmacdope, IAlibay

  * 0.6.0

  Changes

  * Add package deployment via github actions (PR #51 and #50)
  * Re-reformat packages into pyedr and panedr (PR #50)
  * Reformat panedr into two packages: panedr and panedrlite (PR #42)
  * Add GitHub actions workflow for CI (PR #32)
  * fixed setup.cfg for installing from source (PR #28)

  Fixes

  * Fix coverage issues and add coveragerc (PR #47)


13/01/2019 jbarnoud

  * 0.5.2

  Changes

  * Fix typo in \_\_version\_\_
  * Fix typo in \_\_version\_\_


11/01/2019 jbarnoud

  * 0.5.1

  Changes

  * Ignore python3 bytecode in the repo
  * Hopefully allow deploy on tags


11/01/2019 jbarnoud, ezavod

  * 0.5.0

  Changes

  * Add correct handling of blocks and subblocks
  * Fix pypi password for deployment
  * Deploy on the real pypi
  * Only deploy master
  * Fix typo in trove classifiers
  * User correctly encrypted password for deploy
  * Skip file generation by pbr
  * Allow deployment on the pbr branch
  * Try automatic deploy
  * Restore the extra test requirements
  * Move versioning to pbr
  * fix: Initiate an empty frame for each new frame
  * fix: Add correct handling of blocks and subblocks
  * Bump version to 0.4.0
  * Tell travis to use ubuntu xenial
  * Add python 3.7 to travis
  * Fix compatibility issues with python 3.7


20/12/2018 jbarnoud

  * 0.4.0

  Changes

  * Fix a typo in the install, bump to 0.3.1


20/09/2018 jbarnoud, mnmelo, kain88-de

  * 0.3

  Changes

  * Bumb version number to 0.3
  * Tell travis about the move of the tests
  * Move the tests out of the package
  * Travis install pathlib backport on python 2.7
  * Accept \`pathlib.Path\` as an argument for edr\_to\_df
  * Bump minimum python3 version to 3.5
  * Remove pandas \`as\_matrix() deprecated method
  * Update README to refect the use of pytest
  * update python version in ci
  * add manifest
  * Test cleanup
  * Try using py.test instead of pytest
  * Moved to pytest testing
  * Added double-precision support
  * Remove the now unused Enxnms named tuple
  * Move to a more object oriented architecture
  * Fix a typo in the README


15/01/2016 jbarnoud

  * 0.2

  Changes

  * Move to version 0.2
  * Ignore gromacs backup files
  * Do not write an incomplete frame
  * Add test for the progress display in verbose mode
  * Replace the test file for regular dt
  * Add a verbose mode


13/01/2016 jbarnoud

  * 0.1.1

  Changes

  * Move to v0.1.1 and add \_\_version\_\_
  * Update README to account for inclusion in PyPi
  * Make the wheels universal


13/01/2016 jbarnoud

  * 0.1

  Changes

  * Prepare repo for inclusion in PyPi
  * Python 3.2 is not supported
  * Fix errors in .travis.yml
  * Add TravisCI badge in the README
  * Set up TravisCI
  * Add tests
  * Add a gitignore file
  * Fix typo in docstring
  * Adapt example to reflect unicode column names
  * Revert license change of format in the README
  * Update the README
  * Fix unicode issues
  * Panedr is now a proper package with setup.py
  * Add a module docstring
  * Remove the initial notebook
  * Add the licence on top of the library
  * Add a README and a license
  * Translate prototype to a library
  * Initial commit


pyedr/pyedr/pyedr.py
#-*- coding:utf-8 -*-
# PyEDR -- a library to manipulate Gromacs EDR file in python
# Copyright (C) 2022  Jonathan Barnoud
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor,
# Boston, MA  02110-1301  USA

# Most of this file is a python rewrite of part of
# `src/gromacs/fileio/enxio.c` from Gromacs 5.1.
# See gromacs.org.

"""
PyEDR -- Read Gromacs energy file (EDR) in python
==================================================

The ``pyedr`` library allows to read and manipulate the content of Gromacs
energy file (.edr files) in python.

The current version of ``pyedr`` tries to be in par with Gromacs 5.1.1 and
newer when it comes to reading EDR files.

The library exposes the following functions:

- the :func:`read_edr` function parses an EDR file and returns the energy terms
  in a nested list

- the :func:`edr_to_dict` function that turns the nested list created by
  :func:`read_edr` into a dictionary that maps term names to numpy arrays

.. autofunction:: edr_to_dict
"""
from mda_xdrlib import xdrlib
import collections
import warnings
import sys
import itertools
from tqdm import tqdm
from typing import List, Tuple, Dict

import numpy as np


#Index for the IDs of additional blocks in the energy file.
#Blocks can be added without sacrificing backward and forward
#compatibility of the energy files.

#For backward compatibility, the order of these should not be changed.


(enxOR,     # Time and ensemble averaged data for orientation restraints
 enxORI,    # Instantaneous data for orientation restraints
 enxORT,    # Order tensor(s) for orientation restraints
 ensDISRE,  # Distance restraint blocks
 enxDHCOLL, # Data about the free energy blocks in this frame
 enxDHHIST, # BAR histogram
 enxDH,     # BAR raw delta H data
 enxNR      # Total number of extra blocks in the current code,
            # note that the enxio code can read files written by
            # future code which contain more blocks.
) = range(8)

# xdr_datatype
# note that there is no data type 'real' because
# here we deal with the types as they are actually written to disk.
(xdr_datatype_int, xdr_datatype_float, xdr_datatype_double,
 xdr_datatype_int64, xdr_datatype_char, xdr_datatype_string) = range(6)

Enxnm = collections.namedtuple('Enxnm', 'name unit')
ENX_VERSION = 5

__all__ = ['edr_to_dict', 'read_edr', 'get_unit_dictionary']

class EDRFile(object):
    def __init__(self, path):
        with open(path, 'rb') as infile:
            content = infile.read()
        self.data = GMX_Unpacker(content)
        self.do_enxnms()

    def __iter__(self):
        while True:
            try:
                self.frame = Frame()
                self.do_enx()
            except EOFError:
                return
            else:
                yield self.frame

    def do_enxnms(self):
        bReadFirstStep = False
        data = self.data
        magic = data.unpack_int()

        if magic > 0:
            # Assume this is an old edr format
            file_version = 1
            nre = magic
            bOldFileOpen = True
        else:
            bOldFileOpen = False
            if magic != -55555:
                raise ValueError("Energy names magic number mismatch, this is not a GROMACS edr file")
            file_version = ENX_VERSION
            file_version = data.unpack_int()
            if (file_version > ENX_VERSION):
                raise ValueError('Reading file version {} with version {} implementation'.format(file_version, ENX_VERSION))
            nre = data.unpack_int()
        if file_version != ENX_VERSION:
            warnings.warn('Note: enx file_version {}, implementation version {}'.format(file_version, ENX_VERSION))
        nms = edr_strings(data, file_version, nre)

        self.file_version = file_version
        self.nre = nre
        self.nms = nms
        self.bOldFileOpen = bOldFileOpen
        self.bReadFirstStep = False

    def do_eheader(self, nre_test):
        data = self.data
        file_version = self.file_version
        fr = self.frame

        magic = -7777777
        zero = 0
        dum = 0
        tempfix_nr = 0
        ndisre = 0
        startb = 0

        bWrongPrecision = False
        bOK = True

        # We decide now whether we're single- or double-precision. Just peek
        # ahead and see whether we find the magic number where it should.
        base_pos = data.get_position()
        data.set_position(base_pos + 4)
        data.gmx_double = not is_frame_magic(data)
        data.set_position(base_pos)

        first_real_to_check = data.unpack_real()
        if first_real_to_check > -1e-10:
            # Assume we are reading an old format
            file_version = 1
            fr.t = first_real_to_check
            fr.step = data.unpack_int()
        else:
            if not is_frame_magic(data):
                raise ValueError("Energy header magic number mismatch, this is not a GROMACS edr file")
            file_version = data.unpack_int()
            if file_version > ENX_VERSION:
                raise ValueError('Reading file version {} with version {} implementation'.format(file_version, ENX_VERSION))
            fr.t = data.unpack_double()
            fr.step = data.unpack_hyper()
            fr.nsum = data.unpack_int()
            if file_version >= 3:
                fr.nsteps = data.unpack_hyper()
            else:
                fr.nsteps = max(1, fr.nsum)
            if file_version >= 5:
                fr.dt = data.unpack_double()
            else:
                fr.dt = 0
        fr.nre = data.unpack_int()
        if file_version < 4:
            ndisre = data.unpack_int()
        else:
            # now reserved for possible future use
            data.unpack_int()
        fr.nblock = data.unpack_int()
        assert fr.nblock >= 0
        if ndisre != 0:
            if file_version >= 4:
                raise ValueError("Distance restraint blocks in old style in new style file")
            fr.nblock += 1
        # Frames could have nre=0, so we can not rely only on the fr.nre check
        if (nre_test >= 0
            and ((fr.nre > 0 and fr.nre != nre_test)
                 or fr.nre < 0 or ndisre < 0 or fr.nblock < 0)):
            bWrongPrecision = True
            return
        #  we now know what these should be, or we've already bailed out because
        #  of wrong precision
        if file_version == 1 and (fr.t < 0 or fr.t > 1e20 or fr.step < 0):
            raise ValueError("edr file with negative step number or unreasonable time (and without version number).")
        fr.add_blocks(fr.nblock)
        startb = 0
        if ndisre > 0:
            # sub[0] is the instantaneous data, sub[1] is time averaged
            fr.block[0].add_subblocks(2)
            fr.block[0].id = enxDISRE
            fr.block[0].sub[0].nr = ndisre
            fr.block[0].sub[1].nr = ndisre
            fr.block[0].sub[0].type = dtreal
            fr.block[0].sub[1].type = dtreal
            startb += 1
        # read block header info
        for b in range(startb, fr.nblock):
            if file_version < 4:
                # blocks in old version files always have 1 subblock that
                # consists of reals.
                fr.block[b].add_subblocks(1)
                nrint = data.unpack_int()
                fr.block[b].id = b - startb
                fr.block[b].sub[0].nr = nrint
                fr.block[b].sub[0].typr = dtreal
            else:
                fr.block[b].id = data.unpack_int()
                nsub = data.unpack_int()
                fr.block[b].nsub = nsub
                fr.block[b].add_subblocks(nsub)
                for sub in fr.block[b].sub:
                    typenr = data.unpack_int()
                    sub.nr = data.unpack_int()
                    sub.type = typenr
        fr.e_size = data.unpack_int()
        # now reserved for possible future use
        data.unpack_int()
        data.unpack_int()

        # here, stuff about old versions

    def do_enx(self):
        data = self.data
        fr = self.frame

        file_version = -1
        framenr = 0
        frametime = 0
        try:
            self.do_eheader(-1)
        except ValueError:
            print("Last energy frame read {} time {:8.3f}".format(framenr - 1,
                                                                  frametime))
            raise RuntimeError()
        framenr += 1
        frametime = fr.t

        bSane = (fr.nre > 0)
        for block in fr.block:
            bSane |= (block.nsub > 0)
        if not (fr.step >= 0 and bSane):
            raise ValueError('Something went wrong')
        if fr.nre > fr.e_alloc:
            for i in range(fr.nre - fr.e_alloc):
                fr.ener.append(Energy(0, 0, 0))
            fr.e_alloc = fr.nre
        for i in range(fr.nre):
            fr.ener[i].e = data.unpack_real()
            if file_version == 1 or fr.nsum > 0:
                fr.ener[i].eav = data.unpack_real()
                fr.ener[i].esum = data.unpack_real()
                if file_version == 1:
                    # Old, unused real
                    data.unpack_real()

        # Old version stuff to add later

        # Read the blocks
        ndo_readers = (ndo_int, ndo_float, ndo_double,
                       ndo_int64, ndo_char, ndo_string)
        for block in fr.block:
            for sub in block.sub:
                try:
                    sub.val = ndo_readers[sub.type](data, sub.nr)
                except IndexError:
                    raise ValueError("Reading unknown block data type: this file is corrupted or from the future")



class Energy(object):
    __slot__ = ['e', 'eav', 'esum']

    def __init__(self, e=0, eav=0, esum=0):
        self.e = 0
        self.eav = 0
        self.esum = 0

    def __repr__(self):
        return '<{} e={}, eav={}, esum={}>'.format(type(self).__name__,
                                                   self.e, self.eav,
                                                   self.esum)

class SubBlock(object):
    def __init__(self):
        self.nr = 0
        self.type = xdr_datatype_float  # should be double
                                        # if compile in double
        self.val = []
        self.val_alloc = 0

    def alloc(self):
        self.val = [0 for _ in range(self.nr)]
        self.vac_alloc = self.nr


class Block(object):
    def __init__(self):
        # See enxblock_init
        self.id = enxOR
        self.nsub = 0
        self.sub = []
        self.nsub_alloc = 0

    def add_subblocks(self, final_number):
        # See add_subblocks_enxblock
        self.nsub = final_number
        if final_number > self.nsub_alloc:
            for _ in range(final_number - self.nsub_alloc):
                self.sub.append(SubBlock())
            self.nsub_alloc = final_number


class Frame(object):
    def __init__(self):
        # See init_enxframe
        self.e_alloc = 0
        self.ener = []
        self.nblock = 0
        self.nblock_alloc = 0
        self.block = []

    def add_blocks(self, final_number):
        # See add_blocks_enxframe
        self.nblock = final_number
        if final_number > self.nblock_alloc:
            for _ in range(final_number - self.nblock_alloc):
                self.block.append(Block())
            self.nblock_alloc = final_number


class GMX_Unpacker(xdrlib.Unpacker):
    """xdrlib.Unpacker subclass that implements `unpack_real`

    Decision on whether to return 32- or 64-bit reals is controlled by the
    `gmx_double` attribute, set to ``False`` by default.
    """
    gmx_double = False

    def unpack_real(self):
        if self.gmx_double:
            return self.unpack_double()
        return self.unpack_float()


def ndo_int(data, n):
    """mimic of gmx_fio_ndo_int in gromacs"""
    return [data.unpack_int() for i in range(n)]


def ndo_float(data, n):
    """mimic of gmx_fio_ndo_float in gromacs"""
    return [data.unpack_float() for i in range(n)]


def ndo_double(data, n):
    """mimic of gmx_fio_ndo_double in gromacs"""
    return [data.unpack_double() for i in range(n)]


def ndo_int64(data, n):
    """mimic of gmx_fio_ndo_int64 in gromacs"""
    return [data.unpack_huge() for i in range(n)]


def ndo_char(data, n):
    """mimic of gmx_fio_ndo_char in gromacs"""
    return [data.unpack_char() for i in range(n)]


def ndo_string(data, n):
    """mimic of gmx_fio_ndo_string in gromacs"""
    return [data.unpack_string() for i in range(n)]


def edr_strings(data, file_version, n):
    nms = []
    for i in range(n):
        name = data.unpack_string().decode('ascii')
        if file_version >= 2:
            unit = data.unpack_string().decode('ascii')
        else:
            unit = 'kJ/mol'
        nms.append(Enxnm(name=name, unit=unit))
    return nms


def is_frame_magic(data):
    """Unpacks an int and checks whether it matches the EDR frame magic number

    Does not roll the reading position back.
    """
    magic = data.unpack_int()
    return magic == -7777777


all_energies_type = List[List[float]]
all_names_type = List[str]
times_type = List[float]
read_edr_return_type = Tuple[all_energies_type,
                             all_names_type,
                             times_type]


def read_edr(path: str, verbose: bool = False) -> read_edr_return_type:
    """Parse EDR files and make contents available in Python

    :func:`read_edr` does the actual reading of EDR files. It is called by
    :func:`edr_to_df` and :func:`edr_to_dict` to provide the file contents.
    Under the hood, it is using :class:`xdrlib.Unpacker` to access the binary
    EDR file.

    Parameters
    ----------
    path : str
        path to EDR file to be read
    verbose : bool
        Optionally show verbose output while reading the file

    Returns
    -------
    all_energies: list[list[float]]
        A nested containing the energy values for each frame found in the EDR
        file
    all_names: list[str]
        A list containing the names of the energy terms found in the file
    times: list[float]
        A list containing the time of each step/frame.
    """
    edr_file = EDRFile(str(path))
    all_energies = []
    all_names = [u'Time'] + [nm.name for nm in edr_file.nms]
    times = []
    for ifr, frame in tqdm(enumerate(edr_file), disable=(not verbose)):
        if frame.ener:
            # Export only frames that contain energies
            times.append(frame.t)
            all_energies.append([frame.t] + [ener.e for ener in frame.ener])
    return all_energies, all_names, times


def get_unit_dictionary(path: str) -> Dict[str, str]:
    """Creates an EDRFile object which executes the :func:`do_enxnms`
    method. This reads the names and units of the EDR data, which is returned
    as a dictionary mapping column names (str) to unit names (str).

    Parameters
    ----------
    path : str
        path to EDR file to be read

    Returns
    -------
    unit_dict: Dict[str, str]
        A dictionary mapping the term names to their units.
    """
    edr_file = EDRFile(str(path))
    unit_dict = {'Time': "ps"}
    for nm in edr_file.nms:
        unit_dict[nm.name] = nm.unit
    return unit_dict


def edr_to_dict(path: str, verbose: bool = False) -> Dict[str, np.ndarray]:
    """Calls :func:`read_edr` and packs its return values into a dictionary

    The returned dictionary's keys are the names of the energy terms present in
    the EDR file, the values are the time-series energy data for those terms.

    Parameters
    ----------
    path : str
        path to EDR file to be read
    verbose : bool
        Optionally show verbose output while reading the file

    Returns
    -------
    enery_dict: dict[str, np.ndarray]
        dictionary that holds all energy terms found in the EDR file.
    """
    all_energies, all_names, times = read_edr(path, verbose=verbose)
    energy_dict = {}
    for idx, name in enumerate(all_names):
        energy_dict[name] = np.array(
            [all_energies[frame][idx] for frame in range(len(times))])
    return energy_dict
